src/Proxy.hs

```diff
@@ -322,40 +322,93 @@ parseSetCookieExpires' :: BS.ByteString -> Maybe UTCTime
-    (Left err, _) -> do
-      $(logError) $ "Internal error accessing GRPC endpoint: " <> Text.pack err
+  -- Run the client and handle any client-related exceptions.
+  (res, updatedCookies) <- liftIO $ fmap (\(x, y) -> (left show x, y)) (runClientWithCookies cookies cfg c) `catch` exHandler
+  -- Process the result.
+  case res of
+    -- A client error occurred.
+    Left clientError -> do
+      $(logError) $ "Internal error accessing GRPC endpoint: " <> Text.pack (show clientError)
       i <- internationalize
       sendResponseStatus badGateway502 $ object [
         "errorMessage" .= i18n i EMGRPCError,
         "error" .= fromEnum InternalError
         ]
-    (Right (Left err), _) -> do
-      $(logError) $ "GRPC call failed: " <> Text.pack err
-      i <- internationalize
-      sendResponseStatus badGateway502 $ object [
-        "errorMessage" .= i18n i (EMGRPCErrorResponse err),
-        "error" .= fromEnum RequestInvalid
-        ]
-    (Right (Right (GRPCResponse hds a)), updatedCookies) -> do
-      $(logOther "Trace") $ "Got these response headers: " <> Text.pack (show hds)
-      -- set cookies in response
-      let scs = parseSetCookie' . snd <$> (filter $ ("set-cookie" ==) . fst) hds
-      mapM_ setCookie scs
-      $(logOther "Trace") $ "Set-cookies headers to be included in yesod response to client: " <> Text.pack (show scs)
-      -- update cookie map
-      cacheSet $! Map.union updatedCookies cookies
-      k a
+    -- Otherwise the HTTP/2 request succeeded and there is a GRPC result.
+    Right r -> do
+      case r of
+        StatusOk (GRPCResponse hds resultValue) -> do
+          case resultValue of
+            Left err -> do
+              $(logError) $ "No result was available: " <> Text.pack err
+              i <- internationalize
+              sendResponseStatus internalServerError500 $ object [
+                "errorMessage" .= i18n i (EMErrorResponse $ Yesod.InternalError "Error processing the response payload."),
+                "error" .= fromEnum InternalError
+                ]
+            Right val -> do
+              $(logOther "Trace") $ "Got these response headers: " <> Text.pack (show hds)
+              -- set cookies in response
+              let scs = parseSetCookie' . snd <$> (filter $ ("set-cookie" ==) . fst) hds
+              mapM_ setCookie scs
+              $(logOther "Trace") $ "Set-cookies headers to be included in yesod response to client: " <> Text.pack (show scs)
+              -- update cookie map
+              cacheSet $! Map.union updatedCookies cookies
+              k val
+        StatusInvalid -> do
+          $(logError) "Got invalid GRPC status code."
+          i <- internationalize
+          sendResponseStatus badGateway502 $ object [
+            "errorMessage" .= i18n i (EMGRPCErrorResponse "Invalid GRPC status code."),
+            "error" .= fromEnum InternalError
+            ]
```

Comment:
I am not sure, since I believe that invalid GRPC status code refers to an invalid status code being returned in the response to the GRPC call (they are integers in the range [0,..,16]).

Reply:
I guess it should perhaps be a 502, since the response is technically invalid, is it not?